pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+the CPython code base that are changed as a consequence of this PEP.
+So far we have identified around 70 writes spread across a dozen files.
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box".
+
+C extensions that perform writes to data which can be made immutable
+should add the ``Py_CHECKWRITE`` macro shown above on all paths in the
+code that lead to writes to that data.
+
+**TODO**
+  If we don't have a way to tell whether or not an extension 
+  supports immutability, then freeze might end up "leaking"
+  mutable data. Let's discuss this.
+
+We are discussing adding a ``__freeze__`` hook that will be called 
+after an object has been made immutable. This hook could then be used 
+to freeze or otherwise manage any other state on the side that is
+introduced through a C-extension.
+
+
+Freezing Type Which are Not Immutability-Aware
+----------------------------------------------
+
+**TODO** Decide what happens when we attempt to freeze an
+object and discover an object which is not immutability-aware.
+
+1. Throw an exception
+2. Leave it and live with the unsoundness
+3. Nullify the reference to the non immutability-aware object
+4. ...
+
+Somewhere we should also talk about what happens if we throw
+an error during freezing. Do we leave some structures partially
+frozen? If we don't want that, we need to save a log of things
+to unfreeze in case of an error.
+
+
+Examples of Uses of CHECKWRITE
+------------------------------
+
+Inspiration and examples can be found by looking at existing
+uses of ``Py_CHECKWRITE`` in the CPython codebase. Two good
+starting places are ``object.c`` and ``dictobject.c``.
+
+
+Deep Freezing Semantics
+-----------------------
+
+The ``freeze(obj)`` builtin function works as follows:
+
+1. It recursively marks ``obj`` and all objects reachable from ``obj``
+   immutable.
+2. If ``obj`` is already immutable (e.g., an integer, string, or a
+   previously frozen object), it is a no-op.
+3. The freeze operation follows object references (relying on ``tp_traverse`` 
+   in the type structs of the objects involved), including:
+
+    * Object attributes (``__dict__`` for user-defined objects,
+      ``tp_dict`` for built-in types).
+    * Container elements (e.g., lists, tuples, dictionaries,
+      sets).
+    * The ``__class__`` attribute of an object (which makes freezing
+      instances of user-defined classes also freeze their class
+      and its attributes).
+    * The ``__bases__`` chain in classes (freezing a class freezes its
+      base classes).
+
+5. Attempting to mutate a frozen object raises an exception (``MutationError``).
+
+
+Default (Im)Mutabiliy 
+---------------------
+
+- TODO: describe what classes create immutable objects
+
+- Strings, numbers, and tuples of immutable objects create immutable
```

Comment:
only interned strings are marked as immutable at present.

Reply:
Thanks! But also let's discuss this!

Reply:
@matajoh I fail to see where numbers that are dynamically allocated are mutated in-place! Do you have a pointer for this?